commons/src/main/java/com/connexta/replication/data/QueryResponseImpl.java

```diff
@@ -15,18 +15,73 @@
 
 import com.connexta.replication.api.data.Metadata;
 import com.connexta.replication.api.data.QueryResponse;
+import java.util.Iterator;
+import java.util.function.Function;
 
 /** Simple implementation of {@link QueryResponse}. */
 public class QueryResponseImpl implements QueryResponse {
 
   private final Iterable<Metadata> metadata;
 
-  public QueryResponseImpl(Iterable<Metadata> metadata) {
+  private final Function<Exception, RuntimeException> exceptionHandler;
+
+  /**
+   * Instantiates a new query response.
+   *
+   * @param metadata the metadata to provide as part of this response
+   * @param exceptionHandler a exception handler which will be consulted whenever an exception
+   *     occurs while iterating the metadata capable of transforming the exception if needed
+   */
+  public QueryResponseImpl(
+      Iterable<Metadata> metadata, Function<Exception, RuntimeException> exceptionHandler) {
     this.metadata = metadata;
+    this.exceptionHandler = exceptionHandler;
+  }
+
+  private Iterator<Metadata> iterator() {
+    return new Iterator<>() {
+      final Iterator<Metadata> iterator = metadata.iterator();
+
+      @Override
+      public Metadata next() {
+        if (Thread.interrupted()) {
```

Comment:
❓ If a caller calls hasNext before being interrupted, should we let them have the next element to let them finish what they were doing?

Reply:
What does this give us more? 

I only see that it might take a lot longer for them before they detect the interruption and handle it (assuming the processing of that item takes a long time). Typically, if you want to finish off, then you get interrupted, remember you were and go back to processing which wouldn't throw the interruption (unless somebody interrupts the thread again). You choose what you want to do once you are interrupted.